File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> res = new LinkedList<>();
    public List<Integer> preorderTraversal(TreeNode root) {
        TreeNode current = root; 
        TreeNode iPre = null; 
        while(current!=null){
            if(current.left!=null){
                iPre = current.left;
                while(iPre.right!=null){
                    iPre=iPre.right;
                }
                iPre.right = current.right;
                current.right = current.left;
                res.add(current.val);
                current = current.right;
            }
            else{
                res.add(current.val);
                current = current.right;
            }
        }
        return res;
    }
}

File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        if root==None:
            return []
        else:
            left = self.preorderTraversal(root.left)
            right =self.preorderTraversal(root.right)
            
            res = [root.val,*left,*right]
            return res